lib/vdsm/virt/thinp.py

```diff
@@ -553,24 +592,23 @@ def _extend_volume_completed(self, volInfo):
                 with clock.run("refresh-destination-volume"):
                     self._vm.refresh_destination_volume(volInfo)
 
-            with clock.run("refresh-volume"):
-                self._vm.refresh_volume(volInfo)
-
-            # Check if the extension succeeded.  On failure an exception is
-            # raised.
-            # TODO: Report failure to the engine.
-            volSize = self._verify_volume_extension(volInfo)
+            # After a volume was refreshed, the montior thread may wake up and
+            # trigger unwanted extend to the next chunk size. Lock the monitor
+            # during refresh until we set a new threshold.
+            with self._lock:
+                with clock.run("refresh-volume"):
+                    self._vm.refresh_volume(volInfo)
+                volSize = self._verify_volume_extension(volInfo)
 
-            # This was a volume extension or replica and volume extension.
-            clock.stop("total")
-            self._log.info(
-                "Extend volume %s completed %s", volInfo["volumeID"], clock)
+                clock.stop("total")
+                self._log.info(
+                    "Extend volume %s completed %s",
+                    volInfo["volumeID"], clock)
 
-            # Only update apparentsize and truesize if we've resized the leaf
-            if not volInfo['internal']:
-                drive = lookup.drive_by_name(
-                    self._vm.getDiskDevices()[:], volInfo['name'])
-                self.update_drive_volume_size(drive, volSize)
+                if not volInfo['internal']:
+                    drive = lookup.drive_by_name(
+                        self._vm.getDiskDevices()[:], volInfo['name'])
```

Comment:
Just curious, but why slicing here?

Reply:
Ah yeah, I see. Then I would use `list.copy()` instead, is a lot more readable syntax for that purpose.

But is existing code, so never mind. Thanks for clarifying.

Reply:
list.copy() is not the same as [:]. Slicing works with any sequence that may not have a copy method. But the right way to improve this to to move this code into the VM class, it should really have a lookup_disk(name) doing proper locking or copy so caller do not need to deal with these issues.